src/library-aux/scala/Any.scala

```diff
@@ -138,15 +138,107 @@ abstract class Any {
+   *  tells the compiler to forget any type information it has about the receiver object and treat it as if it
+   *  had type `T0`. How the compiler manages to transition from the type of the receiver object to `T0` is 
+   *  **undefined**.
+   *
+   *  Since the outcome is undefined, it is an implementation detail of the compiler backend whether this call 
+   *  results in an operation in the target bytecode (e.g. Java bytecode) or not. 
+   *  Some backends might emit a conversion, a runtime check or replace the receiver object with a default value.
+   *  The decision to do so is mainly driven by the requirements of the target platform and its version.
+   *  There is **absolutely no guarantee** that a certain implementation stays the same between different compiler 
+   *  versions, especially regarding conversions, runtime checks and default values.
+   *
+   *  For instance, if the target is the JVM with version JDK8, then a conversion or a runtime check is added to
+   *  the bytecode in most cases to comply with the java bytecode requirements. However, in some cases also a 
+   *  *default value might be used instead which in turn could be ignored in the context of literal types.
+   *  If a conversion or a runtime check in the form of a
+   *  [checkcast](https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-6.html#jvms-6.5.checkcast) is added,
+   *  then the outcome depends on the implementation of the JVM which should adhere to the rules as in the spec 
+   *  defined in [JLS chapter 5](https://docs.oracle.com/javase/specs/jls/se8/html/jls-5.html). 
+   *  This also means that widening and narrowing conversions take place for primitive types and that the scala 
+   *  compiler has to pass erased types to `checkcast`. Furthermore, neither intersection nor union types nor literal
+   *  types are supported in java bytecode. Whether the Scala compiler backend calculates a lowest upper bound in 
+   *  such a case and passes it  to `checkcast` or `java.lang.Object` is used for simplicity reasons is again an 
+   *  implementation detail.
+   *
+   *  It should be clear from that definition that relying on `asInstanceOf` to do type checking is a very bad idea.
+   *  Use [isInstanceOf] instead for type testing and act as desired if the outcome is `false` (e.g. return a 
+   *  [scala.util.Failure], throw an Exception etc.)
+   *
+   *  Following some examples targeting the JVM and JDK8 using Scala 3.4.0. We start off with statements where the
+   *  outcome is most likely as you would expect and continue with more bizarre outcomes -- hoping that you really 
+   *  refrain from using it as type testing utility, use [isInstanceOf] instead.
+   *
+   *  {{{
+   *  // doesn't insert any operation 1 is statically known to be a subtype of Int
+   *  1.asInstanceOf[Int]
+   *  // inserts a narrowing primitive conversion from double to int
+   *  1.0.asInstanceOf[Int]
```

Comment:
So these are actually de facto specified, although virtually no one knows about them. They are super weird, because they don't follow the general contract in an LSP way. They're almost overloads. A section on what happens when *both* the receiver's static type and the target are primitive types could be added, but otherwise it should be left out of the main "it's undefined" example list. It's an awful inheritance from Java, which also conflated *primitive conversions* and *casts*.

Reply:
I don't mind to leave it out from the examples but I would keep it mentioned that a compiler backend is free to insert conversions in case the subtype relation does not hold

Reply:
But it's *not* free to insert conversions. There are cases where a conversion is mandated, and obeyed by all back-ends that correctly implement the language. And then the rest are *casts*, for which no conversion can happen. There is no room for maybe-a-conversion. There is only room for maybe-a-check.

Reply:
Do we specify how a backend has to implement asInstanceOf? If so, were do I find it?

If we define that the outcome is undefined, then a backend could as well just throw a runtime error instead of converting. Maybe we decide to do so in an upcoming version of scala because, one can use the `toByte`, `to...` methods for conversion.

Reply:
So the way it works is as follows.

Given an expression `x.asInstanceOf[T]` where `x` has the *static* type `S`:

* If `T` is `Unit`, perform a *unit conversion*:
  * Behave like `{ x; () }`, i.e., evaluate `x` then drop the result.
* If `S` and `T` are both primitive numeric types¹, perform a *numeric conversion*:
  * Behave like `x.toT` (`x.toChar`, `x.toByte`, etc.)
* Otherwise, perform a *cast*:
  * If `x eq null`, return the *default value* of type `T`²
  * Otherwise,
    * If `x` is, at run-time, a value of type `T`, return `x`
    * Otherwise, this is a failing cast and the behavior is *undefined*

As you can see, only the very last step of this algorithm leaves any room for interpretation. Back-end-specific behavior can only manifest at that step. All the other "code paths" of that spec are completely well-defined.

¹ a primitive numeric type is one of `Char`, `Byte`, `Short`, `Int`, `Long`, `Float`, `Double`
² the default values are specified in the table [here](https://www.scala-lang.org/files/archive/spec/3.4/04-basic-definitions.html#variable-definitions)

Reply:
thanks, that's nice to know. Now the question is if it is specified like that in which case we could also include it in the docs or whether we consider it an implementation detail which might as well change in the future without previous notice

Reply:
What I wrote above won't change without previous notice. (although I still have hope to warn for the first two top-level bullets, as they have better replacements)

Reply:
good, going to include it in the documentation then

Reply:
@sjrd should I open a bug report because of the following behaviour? According to your rules, the output should be 0 and not `null`
```
val twoFromJava: java.lang.Integer = null
// doesn't insert any operation, outputs `null`
println(twoFromJava.asInstanceOf[2])
```

Reply:
I think that behavior is per spec, expected type is `Any` for println.

`(null: Integer): Int` or `null.asInstanceOf[42]` yield zero (the default value) via unboxing or definition of `null.asInstanceOf`.

Reply:
`println(null.asInstanceOf[2])` outputs 0 its just the case if null is in a variable of a boxed type (java.lang.Integer/Double etc.). IMO it would be fine to say, this falls in the category "the outcome is undefined" because the subtype relation does not hold. But if we want to include the rule in the doc, then I guess it should be fixed.
The longer I think about it the more I think we should not include the rule in the doc and just keep it simpler, something like `This operation is considered safe if the subtype relation holds and is undefined otherwise`
This would allow us to align asInstanceOf closer to isInstanceOf. Especially in the following cases:
```
 x.asInstanceOf[Unit] => throw instead of returning ()
 1.0.asInstanceOf[Int] => throw instead of adding a conversion
 x.asInstanceOf[ Y | Z ] => rewrite to instanceOf Y || instanceOf Z and throw in case it is false instead of calculating the lowest upper bound
 x.asInstanceOf[ Y & Z] = rewrite to instanceOf Y && instanceOf Z and throw
 ```
I now, we would break a lot of code if we do this. I would suggest we start off with warnings in scala 3 (I would keep scala 2 as is, IMO this ship has sailed) including suggestions for alternatives and then in scala 3.7 or such start to throw. The last point might be more controversial as we would break type tagging scala 2 style but people could switch to opaque types right?

Reply:
There is zero chance we're going to do
```scala
 x.asInstanceOf[ Y | Z ] => rewrite to instanceOf Y || instanceOf Z and throw in case it is false instead of calculating the lowest upper bound
 x.asInstanceOf[ Y & Z] = rewrite to instanceOf Y && instanceOf Z and throw
```
That would make `asInstanceOf` slower for the sake of making it every so slightly less undefined. Remember that your `Y | Z` might be hidden behind a type variable `T` for which you have zero knowledge, so it will still be broken even for that "use case".

Reply:
You are right about the hidden aspect, something we should mention in isInstanceOf